frontend/src/metabase/parameters/utils/date-formatting.ts

```diff
@@ -111,52 +114,68 @@ function getFilterTitle(filter: any[]) {
   return prefix + desc;
 }
 
-export function formatAllOptionsWidget(urlEncoded: string) {
+export function formatAllOptionsWidget(urlEncoded: string): string | null {
   if (urlEncoded == null) {
     return null;
   }
   const filter = dateParameterValueToMBQL(urlEncoded, noopRef);
-
   return filter ? getFilterTitle(filter) : null;
 }
 
 function parseDateRangeValue(value: string) {
   const [start, end] = (value || "").split(RANGE_SEPARATOR);
-  return { start, end };
+  return { start: moment(start, true), end: moment(end, true) };
 }
 
-export function formatRangeWidget(value: string) {
+export function formatRangeWidget(value: string): string | null {
   const { start, end } = parseDateRangeValue(value);
-  return start && end
-    ? moment(start).format("MMMM D, YYYY") +
-        " - " +
-        moment(end).format("MMMM D, YYYY")
-    : "";
+  return start.isValid() && end.isValid()
+    ? start.format("MMMM D, YYYY") + " - " + end.format("MMMM D, YYYY")
+    : null;
 }
 
-export function formatSingleWidget(value: string) {
-  return value ? moment(value).format("MMMM D, YYYY") : "";
+function formatSingleWidget(value: string): string | null {
+  const m = moment(value, true);
+  return m.isValid() ? m.format("MMMM D, YYYY") : null;
 }
 
-export function formatMonthYearWidget(value: string) {
-  const m = moment(value, "YYYY-MM");
-  return m.isValid() ? m.format("MMMM YYYY") : "";
+function formatMonthYearWidget(value: string): string | null {
+  const m = moment(value, "YYYY-MM", true);
+  return m.isValid() ? m.format("MMMM YYYY") : null;
 }
 
-export function formatQuarterYearWidget(value: string) {
-  const m = moment(value, "[Q]Q-YYYY");
-  return m.isValid() ? m.format("[Q]Q YYYY") : "";
+function formatQuarterYearWidget(value: string): string | null {
+  const m = moment(value, "[Q]Q-YYYY", true);
+  return m.isValid() ? m.format("[Q]Q YYYY") : null;
 }
 
-export function formatRelativeWidget(value: string) {
+function formatRelativeWidget(value: string): string | null {
   return DATE_MBQL_FILTER_MAPPING[value]
     ? DATE_MBQL_FILTER_MAPPING[value].name
-    : "";
+    : null;
+}
+
+export function formatDateValue(
+  value: string,
+  parameter: UiParameter,
+): string | null {
+  // the value can be from a mismatching parameter, so we need to test every date parameter type
+  const types = [
+    parameter.type,
+    ...PARAMETER_OPERATOR_TYPES.date.map(({ type }) => type),
```

Comment:
We should pay attention to the order in `PARAMETER_OPERATOR_TYPES` array.
I think `date/range` should be before `date/single`, otherwise ranges will get treated as single dates.

It'd be great if we had unit tests for `formatDateValue`.

Reply:
With strict parsing the order should matter (unless there is a bug)

Reply:
Tests for these changes are in `formatting.unit.spec.ts`

Reply:
@kamilmielnik added a test for date ranges